### ISSUE TITLE
ChannelWriteStream has a bug closing underlying stream

### ISSUE THREAD

**Issue by bbyk** [vert-x3/vertx-lang-kotlin#256]
### Questions

`ChannelWriteStream` closes the underlying stream too early

### Version

Vert.x 4.4.5

### Context

When you call [ChannelWriteStream#close](https://github.com/vert-x3/vertx-lang-kotlin/blob/78058ca58158ed5d93feaccc26d44d55a7a1e36a/vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ReceiveChannelHandler.kt#L234-L238) the underlying stream gets closed immediately whereas the channel might still have an item or two in it and [the job reading from the channel](https://github.com/vert-x3/vertx-lang-kotlin/blob/78058ca58158ed5d93feaccc26d44d55a7a1e36a/vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ReceiveChannelHandler.kt#L198-L221) has the opportunity to complete them and in fact is still running and might encounter the closed stream

To illustrate the point, consider the following changes:
1.  `ChannelWriteStream#close` just needs to close the channel
```
    override fun close(cause: Throwable?): Boolean {
        return channel.close(cause)
    }
```
2. The reading job should close the stream instead when the channel is closed, all the remaining items are received and you exit the loop. The suggested change is to use try-finally over the original code and close the stream in the finally block.
```
        launch {
            try {
                while (true) {
                    val res = channel.receiveCatching()
                    ....
                }
            } finally {
                stream.end()
            }
        }

```
### Do you have a reproducer?

No, because it's a race condition.

**Comment by tsegismont** [comment]
Please take a look at #260 if you get a chance

**Comment by tsegismont** [comment]
Fixed by cd81bb6

**Comment by bbyk** [comment]
@tsegismont Perfect. Thanks!
